core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java

```diff
@@ -13,86 +13,186 @@
                 .process(expression);
     }
 
     private static class ConnectorToSqlExpressionTranslator
     {
-        private final Map<String, Symbol> variableMappings;
+        private final Session session;
+        private final Metadata metadata;
         private final LiteralEncoder literalEncoder;
+        private final Optional<Map<String, Symbol>> variableMappings;
 
-        public ConnectorToSqlExpressionTranslator(Map<String, Symbol> variableMappings, LiteralEncoder literalEncoder)
+        public ConnectorToSqlExpressionTranslator(Session session, Metadata metadata, LiteralEncoder literalEncoder, Optional<Map<String, Symbol>> variableMappings)
         {
-            this.variableMappings = requireNonNull(variableMappings, "variableMappings is null");
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
             this.literalEncoder = requireNonNull(literalEncoder, "literalEncoder is null");
+            this.variableMappings = requireNonNull(variableMappings, "variableMappings is null");
         }
 
-        public Expression translate(ConnectorExpression expression)
+        public Optional<Expression> process(ConnectorExpression expression)
         {
-            if (expression instanceof Variable) {
-                return variableMappings.get(((Variable) expression).getName()).toSymbolReference();
+            if (expression instanceof Constant) {
+                return Optional.of(literalEncoder.toExpression(((Constant) expression).getValue(), expression.getType()));
+            }
+            else if (expression instanceof FieldDereference) {
+                FieldDereference fieldDereference = (FieldDereference) expression;
+                Optional<Expression> baseOpt = process(fieldDereference.getTarget());
+                return baseOpt.map(base -> new SubscriptExpression(base, new LongLiteral(Long.toString(fieldDereference.getField() + 1))));
             }
+            else if (expression instanceof Variable) {
+                String name = ((Variable) expression).getName();
+                return Optional.of(variableMappings.map(map -> map.get(name).toSymbolReference())
+                                                   .orElse(new SymbolReference(name)));
+            }
+            else if (expression instanceof Call) {
+                return processCall((Call) expression);
+            }
+            return Optional.empty();
+        }
 
-            if (expression instanceof Constant) {
-                return literalEncoder.toExpression(((Constant) expression).getValue(), expression.getType());
+        protected Optional<Expression> processCall(Call call)
+        {
+            if (LIKE_FUNCTION_NAME.equals(call.getName())) {
+                return translateLike(call.getArguments().get(0), call.getArguments().get(1));
+            }
+
+            QualifiedName qualifiedName = getQualifiedName(call);
```

Comment:
This is use only in one place during construction of `FunctionCall` below. Inline it.

Reply:
Done